src/coreclr/src/vm/frames.cpp

```diff
@@ -454,7 +458,14 @@ VOID Frame::Pop(Thread *pThread)
              (*m_Next->GetGSCookiePtr() == GetProcessGSCookie()));
 
     pThread->SetFrame(m_Next);
-    m_Next = NULL;
+
+    // InlinedCallFrames are initialized in the prolog of a jitted method with PInvokes, so
+    // we must not change the value of m_Next here, otherwise when we link an ICF during a PInvoke
+    // call, we will lose the rest of the frames chain.
+    if (GetVTablePtr() != InlinedCallFrame::GetMethodFrameVPtr())
```

Comment:
This doesn't seem right. I don't understand why the NULL here would matter. The m_Next is set during the Push operation, so if we pop the frame and then push it again, it should work.

Reply:
The loops that walk the chain check typically do `while(pFrame != FRAME_TOP)`, and this setting to NULL breaks that and causes AVs. Note that with my changes, the behavior of pushing/popping ICFs is different than before, and setting this to NULL was never a problem before because the ICF was always in the chain for jitted code. Now we have a state where the ICF is initialized, but only linked/unliked during pinvokes.

These changes are still at the experimental stage anyways. I'm kinda doing them on the side, and testing things out using this PR :)

Reply:
Also we do not use the Push operation in frames.cpp to push InlinedCallFrames (they are pushed either by the JIT helper for R2R code, or by codegen for jitted code)

Reply:
My point was that we should never reach a frame that was popped off the chain when doing stack walk. If we do, then I believe it is something that needs to be fixed.